File: Grind75/62-FindAllAnagramsOfString/python/FindAllAnagramsOfString.py
#!/bin/python3

from __future__ import annotations
import sys
from typing import Counter
sys.path.append('../../Utility/python')
from PrettyPrint import *
from Colors import *

class Solution:
  def findAnagrams(self, s: str, p: str) -> List[int]:
    anagrams: List[int] = []
    if len(s) == 0 or len(s) < len(p):
      return anagrams

    def ind(c: str) -> int: return ord(c) - ord('a')

    pCount = [0] * 26
    for c in p:
      pCount[ind(c)] += 1

    sCount = [0] * 26
    for i in range(len(p) - 1):
      sCount[ind(s[i])] += 1

    for start in range(len(s) - len(p) + 1):
      sCount[ind(s[start + len(p) - 1])] += 1
      if pCount == sCount:
        anagrams.append(start)
      sCount[ind(s[start])] -= 1

    return anagrams


if __name__ == '__main__':
  soln = Solution()
  testCases = [
    ( "abc", "ab", [0] ),
    ( "aba", "ab", [0,1] ),
    ( "bacbacb", "abc", [0,1,2,3,4] ),
    ( "bacdbacb", "abc", [0,4,5] ),
  ]

  for s, p, answer in testCases:
    colorPrint(LIGHT_BLUE, quote(s), ", ", quote(p), " => ", answer, sep='', end='')
    print(' --> ', end='')
    result = soln.findAnagrams(s, p)
    colorPrint(GREEN if result == answer else RED, result)

File: Grind75/62-FindAllAnagramsOfString/cpp/FindAllAnagramsOfString.cpp
#include "FindAllAnagramsOfString.h"
#include <vector>
#include <string>

using namespace std;

vector<int> Solution::findAnagrams(string s, string p) {
  vector<int> anagrams;
  if (s.empty() || s.size() < p.size())
    return anagrams;

  vector<int> pLetters(26, 0);
  for (char c : p)
    pLetters[c - 'a']++;
  vector<int> sLetters(26, 0);
  for (size_t i = 0; i < p.size(); i++)
    sLetters[s[i] - 'a']++;

  for (size_t end = p.size(); end <= s.size(); end++) {
    size_t start = end - p.size();
    if (sLetters == pLetters)
      anagrams.push_back(start);
    if (end < s.size()) {
      sLetters[s[start] - 'a']--;
      sLetters[s[end] - 'a']++;
    }
  }

  return anagrams;
}

